### ISSUE TITLE
Add get-minidump-instructions support to minidump-processor/minidump-stackwalk?

### ISSUE THREAD

**Issue by Gankra** [luser/rust-minidump#458]
https://hg.mozilla.org/users/tmielczarek_mozilla.com/get-minidump-instructions/

https://hg.mozilla.org/users/tmielczarek_mozilla.com/get-minidump-instructions/file/tip/get-minidump-instructions.cc

> retrieves and displays the memory range surrounding the faulting instruction pointer from a minidump. You will almost always want to run it with the --disassemble option, which will make it send the bytes through objdump to display the disassembled instructions. If you also give it a path to symbols (see stackwalk-http above) it can download the matching source files from hg.mozilla.org and display source interleaved with the disassembly.

Specifically the --disassemble feature would be nice to provide. By default I would use [yaxpeax](https://github.com/iximeow/yaxpeax-x86) for this, which is a dependency I'm always *just* about to add and then end up not needing.

From what it sounds like, minidump clients tend to write out the section of the binary that the crash occurred at, so even without pulling the binary down you tend to be able to grab this range of memory from the minidump itself (untested). I believe this is the principle used by breakpad's old [exploitability analysis](https://searchfox.org/mozilla-central/rev/8a0c04ef0242679016a086e56cae5d707a224d34/toolkit/crashreporter/google-breakpad/src/processor/exploitability_linux.cc#218-220).

This issue filed on behalf of jld, who mentioned that he has a script similar to this that he uses with minidump_dump.

**Comment by gabrielesvelto** [comment]
Note: being able to disassemble the crashing instruction would also enable us to compute the correct crashing address when it was a non-canonical value (see bug [1493342](https://bugzilla.mozilla.org/show_bug.cgi?id=1493342)) on the Mozilla tracker.

**Comment by luser** [comment]
I already wrote this, you removed it. :)
https://github.com/luser/rust-minidump/blob/23c3ddccef84b3bc6d36a1d62d7fad746fe12346/minidump-tools/src/lib.rs#L232-L289

In fairness, my [`disasm`](https://github.com/luser/rust-disasm/) crate pulled in some annoying dependencies. I wrote it using `capstone`, but there's [a `yaxpeax` branch](https://github.com/luser/rust-disasm/tree/yaxpeax) that I think is in a working state.

**Comment by luser** [comment]
> From what it sounds like, minidump clients tend to write out the section of the binary that the crash occurred at, so even without pulling the binary down you tend to be able to grab this range of memory from the minidump itself (untested)

Yes, Microsoft's minidump writer attempts to include 256 bytes of memory surrounding the faulting instruction pointer, so I implemented that same behavior for the Mac and Linux minidump writers:
https://searchfox.org/mozilla-central/rev/68a5327697ec43aa55b458c504c4b313c9c80528/toolkit/crashreporter/breakpad-client/linux/minidump_writer/minidump_writer.cc#425
https://searchfox.org/mozilla-central/rev/68a5327697ec43aa55b458c504c4b313c9c80528/toolkit/crashreporter/breakpad-client/mac/handler/minidump_generator.cc#1036

**Comment by Gankra** [comment]
> I already wrote this, you removed it. :)

WHELP

but yeah much like the dwarf support it was a good idea but needed too much TLC. This definitely is a useful pointer to get started though :)

One thing we could consider as a first step is to just expose that raw binary dump to the user in non-dump modes, and then they can disassemble it however they want. 256 bytes is pretty small!

**Comment by Gankra** [comment]
If we *do* include disassembly stuff, depending on how much it bloats the binary we should definitely consider making it an optional feature.

(I still want to believe in the dream of One True Tool because I really hate having to tell people "whelp, now you need to pull up...")

**Comment by luser** [comment]
> One thing we could consider as a first step is to just expose that raw binary dump to the user in non-dump modes, and then they can disassemble it however they want. 256 bytes is pretty small!

It's true! Putting the instruction bytes in a hex string in the output somewhere would be useful, although you'd want to ensure they were not exposed publicly on crash-stats. If you look at the source of `get-minidump-instructions` you can see that passing that to `objdump` is still a bit of a PITA, so having tooling for this is nice.

> (I still want to believe in the dream of One True Tool because I really hate having to tell people "whelp, now you need to pull up...")

I can split the previous implementation out into a separate repo, and set up some CI to build release binaries for major platforms. That would at least make the directions as simple as "download a binary from the releases page here".

**Comment by gabrielesvelto** [comment]
We now have disassembly support so it might be just a matter of implementing the `get-minidump-instructions` functionality at this point, right?

**Comment by luser** [comment]
> We now have disassembly support so it might be just a matter of implementing the `get-minidump-instructions` functionality at this point, right?

Yeah, if the disassembly library in use can serialize instructions out to assembly syntax, a basic implementation of this should be pretty straightforward.

**Comment by mgaudet** [comment]
I am trying to look at an android dump, and this would be helpful -- The version in Ted's repo is not able to process this dump :)

**Comment by gabrielesvelto** [comment]
At this point it's a matter of wiring up [*yaxpeax-arm*](https://github.com/iximeow/yaxpeax-arm) so we get disassembly functionality for ARMv7 and ARMv8. It's worth keeping in mind that the crate is still considered incomplete so it might not have all the functionality of *yaxpeax-x86* just yet.

**Comment by luser** [comment]
On that rust-disasm branch where I tried out yaxpeax I was able to make the implementation generic over architectures and just select the right one using an enum:
https://github.com/luser/rust-disasm/blob/c7ffc6fddcbfabe01c96682cd8933331af83e4f5/src/lib.rs#L277-L315
